File: modules/expo-settings/ios/ExpoSettingsModule.swift
import ExpoModulesCore

public class ExpoSettingsModule: Module {
  let ORIENTATION_EVENT = "onOrientationChange"
  
  public func definition() -> ModuleDefinition {
    Name("ExpoSettings")

    Events(ORIENTATION_EVENT)

    Function("getOrientation") { () -> String in 
      return currentOrientation() 
    }

    OnStartObserving(ORIENTATION_EVENT) {
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(orientationDidChange),
        name: UIDevice.orientationDidChangeNotification,
        object: nil
      )
    }

    OnStopObserving(ORIENTATION_EVENT) {
        NotificationCenter.default.removeObserver(
        self,
        name: UIDevice.orientationDidChangeNotification,
        object: nil
      )
    }

    View(ExpoSettingsView.self) {}
  }

  private func currentOrientation() -> String {
    switch UIDevice.current.orientation {
    case .landscapeLeft, .landscapeRight:
      return "landscape"
    case .portrait, .portraitUpsideDown:
      return "portrait"
    default:      
      let bounds = UIScreen.main.bounds
      return bounds.width > bounds.height ? "landscape" : "portrait"
   }
  }

 @objc private func orientationDidChange() {
    let orientation = currentOrientation()
    sendEvent(ORIENTATION_EVENT, ["orientation": orientation])
  }
}

File: modules/expo-settings/android/src/main/java/expo/modules/settings/ExpoSettingsModule.kt
package expo.modules.settings

import android.content.ComponentCallbacks
import android.content.res.Configuration
import android.content.res.Resources
import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoSettingsModule : Module() {
  companion object {
    private const val ORIENTATION_EVENT = "onOrientationChange"
  }

  private val context
    get() = requireNotNull(appContext.reactContext)

  override fun definition() = ModuleDefinition {
    Name("ExpoSettings")

    Events(ORIENTATION_EVENT)

    Function("getOrientation") {
      return@Function currentOrientation()
    }

    OnStartObserving(ORIENTATION_EVENT) {
      context.applicationContext.registerComponentCallbacks(configCallback)
    }

    OnStopObserving(ORIENTATION_EVENT) {
      context.applicationContext.unregisterComponentCallbacks(configCallback)
    }
  }

  private fun currentOrientation(): String {
    val config = appContext.currentActivity?.resources?.configuration
      ?: Resources.getSystem().configuration
    return orientationString(config.orientation)
  }

  private fun orientationString(orientation: Int) =
    if (orientation == Configuration.ORIENTATION_LANDSCAPE) "landscape" else "portrait"

  private val configCallback = object : ComponentCallbacks {
    override fun onConfigurationChanged(newConfig: Configuration) {
      sendEvent(ORIENTATION_EVENT, bundleOf("orientation" to orientationString(newConfig.orientation)))
    }

    override fun onLowMemory() {}
  }
}
